src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java

```diff
@@ -0,0 +1,300 @@
+         [5]{4}StepAtomNode Print Message
+           -LogActionImpl Console Output
+         [6]{5}StepStartNode Stage : Start
+         [7]{6}StepStartNode second
+           -BodyInvocationAction null
+           -LabelAction second
+         [8]{7}StepAtomNode Print Message
+           -LogActionImpl Console Output
+         [9]{8}StepEndNode Stage : Body : End  [st=7]
+           -BodyInvocationAction null
+         [10]{9}StepEndNode Stage : End  [st=6]
+         [11]{10}StepAtomNode Print Message
+           -LogActionImpl Console Output
+         [12]{11}StepAtomNode third
+           -LogActionImpl Console Output
+           -LabelAction third
+           -StageActionImpl null
+         [13]{12}StepAtomNode Print Message
+           -LogActionImpl Console Output
+         [14]{13}StepStartNode Stage : Start
+         [15]{14}StepStartNode fourth
+           -BodyInvocationAction null
+           -LabelAction fourth
+         [16]{15}StepAtomNode Print Message
+           -LogActionImpl Console Output
+         [17]{16}StepEndNode Stage : Body : End  [st=15]
+           -BodyInvocationAction null
+         [18]{17}StepEndNode Stage : End  [st=14]
+         [19]{18}StepAtomNode Print Message
+           -LogActionImpl Console Output
+         [20]{19}FlowEndNode End of Pipeline  [st=2]
+         */
+
+        ForkScanner scan = new ForkScanner();
+        scan.setup(build.getExecution().getCurrentHeads());
+        CollectingChunkVisitor visitor = new CollectingChunkVisitor();
+        scan.visitSimpleChunks(visitor, new StageChunkFinder());
+        ArrayList<MemoryFlowChunk> stages = visitor.getChunks();
+
+        Assert.assertEquals(4, stages.size());
+        assertChunkBoundary(stages.get(0), 3, 4, 6, 7);
+        assertChunkBoundary(stages.get(1), 6, 7, 9, 10);
+        assertChunkBoundary(stages.get(2), 11, 12, 14, 15);
+        assertChunkBoundary(stages.get(3), 14, 15, 17, 18);
+    }
+
+    /** Single-step stage markers */
+    @Test
+    public void testLegacyStage() throws Exception {
+        WorkflowJob job = jenkinsRule.jenkins.createProject(WorkflowJob.class, "Legacy stage job");
+
+        job.setDefinition(new CpsFlowDefinition("" +
+                "node {" +
+                "   stage ('Build'); " +
+                "   echo ('Building'); " +
+                "   stage ('Test'); " +
+                "   echo ('Testing'); " +
+                "   stage ('Deploy'); " +
+                (hudson.Functions.isWindows() ? "   bat ('rmdir /s/q targs || echo no such dir\\n mkdir targs && echo hello> targs\\\\hello.txt'); "
+                        : "   sh ('rm -rf targs && mkdir targs && echo hello > targs/hello.txt'); "
+                ) +
+                "   archive(includes: 'targs/hello.txt'); " +
+                "   echo ('Deploying'); " +
+                "}"));
+        /**
+         * Node dump follows, format:
+         [ID]{parent,ids} flowNodeClassName stepDisplayName [st=startId if a block node]
```

Comment:
Not sure what you mean?  I think I would need to see at least pseudo-syntax to follow what specifically you're describing here.